compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -225,6 +226,28 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                         }
                     })
                     .collect();
+
+                // For this closure introduce fake reads based on the the new data we wrote in
+                // typeck results. Note that the fake reads are introduced into MIR but not into
+                // the closure operands/ aggregate created.
+                //
+                // It's possible that we are creating a nested closure here and therefore weren't
+                // able to introduce fake reads in our parent, ensure that this an UpvarRef.
+                // Because at some point our parent probably introduced a fake read for this place.
+                debug!("RFC2229 = as_rvalue.rs ExprKind::Closure");
+                // [FIXME] Remove feature flag check
+                // Currently is necessary otherwise some of the old implementation fails
+                if this.hir.tcx().features().capture_disjoint_fields {
+                    for thir_place in fake_reads.into_iter() {
+                        debug!("RFC2229 = as_rvalue.rs {:?}", thir_place);
+                        let fake_read_upvar = this.hir.mirror(thir_place);
+                        let mir_place = unpack!(block = this.as_place(block, fake_read_upvar));
+                        debug!("RFC2229 = as_rvalue pushing following mir_place {:?}", mir_place);
+                        this.cfg.push_fake_read(block, source_info, FakeReadCause::ForLet, mir_place);
```

Comment:
There are two fake read causes. One `ForLet` and one `ForMatch`. We probably should capture the correct one. I don't think it will be too much work to do that, but I think if this works fine for now, we should delay that until we fix bugs/ we get close to getting this to work.